apiserver/facades/client/charms/repositories.go

```diff
@@ -31,7 +31,7 @@ var logger = loggo.GetLogger("juju.apiserver.charms")
 // CharmHubClient represents the methods required of a
 // client to install or upgrade a CharmHub charm.
 type CharmHubClient interface {
-	DownloadAndRead(ctx context.Context, resourceURL *url.URL, archivePath string) (*charm.CharmArchive, error)
+	DownloadAndRead(ctx context.Context, resourceURL *url.URL, archivePath string, pb charmhub.ProgressBar) (*charm.CharmArchive, error)
```

Comment:
Probably not worth it but I wonder whether we could convert the last two params into functional args (WithArchivePath, WithProgressBar). This might simplify the various DownloadXXX methods below.

Reply:
We always need the archive path, we just don't need the progress bar, so I've made that optional.